### PR TITLE
[TVMScript] Add syntax sugar for T.handle and T.match_buffer

### PR DESCRIPTION
**Pull request by shingjan**
This PR intends to add syntax sugar for use of `T.handle` and `T.match_buffer` in the parameter list of `TVMScript`.

Previously with TVMScript:
```
def elementwise(
    a: T.handle,
    b: T.handle,
) -> None:
    A = T.match_buffer(a, (128, 128, 128, 128))
    B = T.match_buffer(b, (128, 128, 128, 128))
    # some computation
```

After this PR, user should be able to write TVMScript just below:

```
# can use kwargs
def elementwise(
    a: T.Buffer(shape=(128, 128, 128, 128), dtype="float32", elem_offset=xxx, ...),
    b:T.Buffer(shape=(128, 128, 128, 128), dtype="float32", elem_offset=xxx, ...),
) -> None:
    # some computation

# cannot use kwargs
def elementwise(
    a: T.Buffer[(128, 128, 128, 128), "float32"],
    b:T.Buffer[(128, 128, 128, 128), "float32"],
) -> None:
    # some computation
T.Buffer[(1, 224, 224, 3), "float32"]
```

cc: @vinx13 @junrushao1994

### CODE REVIEW

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (vinx13) [inline tests/python/unittest/test_tvmscript_syntax_sugar.py:24]: please test against expected results using `assert_structual_equal` and let us know when it is ready for review

File: python/tvm/script/tir/ty.py
Comment (vinx13) [inline python/tvm/script/tir/ty.py:58]: let's also support other arguments of `decl_buffer`:
https://github.com/apache/tvm/blob/main/python/tvm/script/tir/special_stmt.py#L154-L159

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (Hzfengsy) [inline tests/python/unittest/test_tvmscript_syntax_sugar.py:52]: The newline at the end of file

Comment (Hzfengsy) [reply tests/python/unittest/test_tvmscript_syntax_sugar.py:52]: Note that this test file contains no value checks but only makes sure the sugar can be parsed. Please add `tests` the parsed result by `assert_structual_equal` or direct check the function `match_buffer`

**Review by Hzfengsy — CHANGES_REQUESTED** [review]

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (shingjan) [reply tests/python/unittest/test_tvmscript_syntax_sugar.py]: done

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:1110]: Though the only usage is match buffer, the visitor here is generic and shouldn’t mention match buffer in error message.

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:1140]: Ditto

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (Hzfengsy) [inline tests/python/unittest/test_tvmscript_syntax_sugar.py:52]: Remove it?

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (Hzfengsy) [inline tests/python/unittest/test_tvmscript_syntax_sugar.py:55]: What's the difference between `Buffer[]` and `Buffer()`?

**Review by Hzfengsy — CHANGES_REQUESTED** [review]

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (vinx13) [inline tests/python/unittest/test_tvmscript_syntax_sugar.py:21]: not needed

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:448]: if this if statement only works for `T.match_buffer` it would be great to add some assertions (otherwise naming the return value `buf` here is confusing)

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:1068]: this function is very similar to `parse_arg_list` is it possible to reuse it (and extend it with needed functionality)

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:1074]: would be better to recursively transform elements of `TypeTuple` in `transform_TypeTuple` than do it here, you can find examples in `transform_Tuple`

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:1083]: ditto

**Review by vinx13 — CHANGES_REQUESTED** [review]

**Comment by junrushao1994** [comment]
What's the status of this PR? Shall we release a minor/patch version of synr and get this in?

**Comment by shingjan** [comment]
@junrushao1994 My change to synr is in. Talking to Tristan about updating it to a newer version so this could in.

File: python/tvm/script/parser.py
Comment (shingjan) [reply python/tvm/script/parser.py:1110]: done

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (shingjan) [reply tests/python/unittest/test_tvmscript_syntax_sugar.py:55]: They are two ways of utilizing the match buffer syntax sugar. `Buffer[]` is for unnamed positional arguments and `Buffer()` is for named keywords arguments.

File: python/tvm/script/parser.py
Comment (shingjan) [reply python/tvm/script/parser.py]: fixed.

**Comment by vinx13** [comment]
please address CI issues

File: python/tvm/script/tir/ty.py
Comment (vinx13) [inline python/tvm/script/tir/ty.py:129]: can you comment the usage of this one? how does it relate to `__call__`? From the current usage like `T.Buffer[...]` I think this should be defined as `__getitem__(self, args)` instead of static method as `Buffer` is an instance of `GenericBufferType`

Comment (shingjan) [reply python/tvm/script/tir/ty.py]: done

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (vinx13) [reply tests/python/unittest/test_tvmscript_syntax_sugar.py:53]: what happened if users don't specify buffer name `a` here, e.g. a: T.Buffer[(128, 128, 128, 128), "float32"]

Comment (shingjan) [reply tests/python/unittest/test_tvmscript_syntax_sugar.py:53]: Those two buffers will have the default name "Buffer" and "Buffer_1"

**Review by vinx13 — APPROVED** [review]

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (shingjan) [reply tests/python/unittest/test_tvmscript_syntax_sugar.py:53]: @Hzfengsy can you take a look of the latest change and approve so we can merge this PR in? Thanks!

**Comment by junrushao1994** [comment]
Let's get this in quickly, and improve the printer in a separate PR

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (junrushao1994) [reply tests/python/unittest/test_tvmscript_syntax_sugar.py:53]: I will second Wuwei on the buffer name. Let's fix it before merging this PR

Comment (Hzfengsy) [reply tests/python/unittest/test_tvmscript_syntax_sugar.py:53]: I suggest the buffer name `a` in case `a: T.Buffer[(128, 128, 128, 128), "float32"]`. Also, we'd better use name `a_handle` for the func param

**Comment by shingjan** [comment]
@junrushao1994 This PR still awaits @Hzfengsy 's approval and I don't seem to have the right to dismiss his request of change.

**Comment by vinx13** [comment]
@shingjan let's fix buffer name and then get this in

**Comment by junrushao1994** [comment]
To be clear, it's important to make the buffer name correct and it's not addressed yet

**Comment by shingjan** [comment]
@junrushao1994 @vinx13 my bad. Overlooked our conversation about buffer name above. Fixing it now.

File: tests/python/unittest/test_tvmscript_syntax_sugar.py
Comment (shingjan) [reply tests/python/unittest/test_tvmscript_syntax_sugar.py]: this is addressed in the latest change

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:465]: since this is not used in match buffer case, move it to `else` branch

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:290]: Rather than checking `len(args)` here we should use more explicit and clear way to check whether buffer name is present. One possible way is to force users to use kwargs for parameters other than `shape` and `dtype` so that you can check and set kw_args here

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:477]: `self.transform(arg.ty)` and `parse_type` is kind of duplication, dtype of `te.var` here is actually the same as `buffer.dtype`

**Review by vinx13 — CHANGES_REQUESTED** [review]

File: python/tvm/script/parser.py
Comment (shingjan) [reply python/tvm/script/parser.py:465]: This is actually used in match buffer syntax sugar case as the `arg.name` needs to later fetched from `self.context.func_params` to be part of match_buffer's argument.

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:298]: remove this

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:1162]: since this method is now specifically for match buffer, you can assert this and remove the error reporting below

**Review by vinx13 — CHANGES_REQUESTED** [review]

File: python/tvm/script/parser.py
Comment (vinx13) [inline python/tvm/script/parser.py:1158]: let's make it more clear, e.g
```
This method is for buffer declarations in the function parameters.
```

Comment (shingjan) [reply python/tvm/script/parser.py]: done

**Review by vinx13 — APPROVED** [review]

**Comment by vinx13** [comment]
@shingjan please rebase

**Comment by shingjan** [comment]
@vinx13 rebased

**Comment by shingjan** [comment]
@Hzfengsy can you take a quick look of the recent change? Thanks!

**Review by Hzfengsy — APPROVED** [review]
Thanks @shingjan

**Pull request merged**
